File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/AbilityUtils.java
package io.github.kdy05.physicalFighters.ability;

import io.github.kdy05.physicalFighters.util.AttributeUtils;
import org.bukkit.Bukkit;
import org.bukkit.FluidCollisionMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.Scoreboard;
import org.bukkit.scoreboard.Team;
import org.bukkit.util.RayTraceResult;

import java.util.function.Consumer;
import java.util.function.Predicate;

public final class AbilityUtils {
    // 인스턴스화 방지용
    private AbilityUtils() {
        throw new AssertionError();
    }

    public static Ability findAbility(Player p) {
        return AbilityRegistry.findAbility(p);
    }

    public static Location getTargetLocation(Player p, int bound) {
        RayTraceResult result = p.getWorld().rayTraceBlocks(
                p.getEyeLocation(), p.getEyeLocation().getDirection(),
                bound, FluidCollisionMode.ALWAYS);
        return result != null && result.getHitBlock() != null
                ? result.getHitBlock().getLocation() : null;
    }

    public static void goVelocity(LivingEntity entity, Location target, double value) {
        entity.setVelocity(entity.getVelocity().add(target.toVector()
                        .subtract(entity.getLocation().toVector()).normalize()
                        .multiply(value)));
    }

    public static void piercingDamage(LivingEntity entity, double damage) {
        entity.setHealth(Math.max(0, entity.getHealth() - damage));
    }

    public static void healEntity(LivingEntity entity, double amount) {
        double maxHealthValue = AttributeUtils.getMaxHealth(entity);
        entity.setHealth(Math.min(maxHealthValue, entity.getHealth() + amount));
    }

    public static void createBox(Location center, Material material, int radius, int height) {
        createBox(center, material, radius, height, false);
    }

    public static void createBox(Location center, Material material, int radius, int height, boolean ignoreBedrock) {
        World world = center.getWorld();
        if (world == null) return;
        for (int y = 1; y <= height; y++) {
            for (int x = -radius; x <= radius; x++) {
                for (int z = -radius; z <= radius; z++) {
                    Location blockLoc = center.clone().add(x, y, z);
                    Block block = world.getBlockAt(blockLoc);
                    if (ignoreBedrock || block.getType() != Material.BEDROCK) {
                        block.setType(material);
                    }
                }
            }
        }
    }

    public static void splashDamage(Player caster, Location location, double bound, double damage) {
        caster.getWorld().getNearbyEntities(location, bound, bound, bound).stream()
                .filter(entity -> entity instanceof LivingEntity)
                .map(entity -> (LivingEntity) entity)
                .filter(entity -> entity != caster)
                .filter(entity -> isNotSameTeam(caster, entity))
                .forEach(entity -> entity.damage(damage, caster));
    }

    public static void splashTask(Player caster, Location location, double bound, Consumer<LivingEntity> action) {
        splashTask(caster, location, bound, entity -> true, action);
    }

    public static void splashTask(Player caster, Location location, double bound,
                                  Predicate<LivingEntity> filter, Consumer<LivingEntity> action) {
        caster.getWorld().getNearbyEntities(location, bound, bound, bound).stream()
                .filter(entity -> entity instanceof LivingEntity)
                .map(entity -> (LivingEntity) entity)
                .filter(entity -> entity != caster)
                .filter(entity -> isNotSameTeam(caster, entity))
                .filter(filter)
                .forEach(action);
    }

    private static boolean isNotSameTeam(Player caster, LivingEntity target) {
        if (!(target instanceof Player)) {
            return true;
        }
        Player targetPlayer = (Player) target;

        if (Bukkit.getScoreboardManager() == null) {
            return true;
        }
        Scoreboard scoreboard = Bukkit.getScoreboardManager().getMainScoreboard();

        Team casterTeam = scoreboard.getEntryTeam(caster.getName());
        Team targetTeam = scoreboard.getEntryTeam(targetPlayer.getName());
        if (casterTeam == null || targetTeam == null) {
            return true;
        }
        return !casterTeam.equals(targetTeam);
    }

}

File: plugin/src/main/kotlin/io/github/kdy05/physicalFighters/ability/AbilityUtils.kt
package io.github.kdy05.physicalFighters.ability

import io.github.kdy05.physicalFighters.PhysicalFighters
import io.github.kdy05.physicalFighters.util.AttributeUtils
import org.bukkit.Bukkit
import org.bukkit.FluidCollisionMode
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable
import java.util.function.Consumer
import java.util.function.Predicate
import kotlin.math.cos
import kotlin.math.sin

object AbilityUtils {

    @JvmStatic
    fun findAbility(p: Player): Ability? = AbilityRegistry.findAbility(p)

    @JvmStatic
    fun getTargetLocation(p: Player, bound: Int): Location? {
        val result = p.world.rayTraceBlocks(
            p.eyeLocation, p.eyeLocation.direction,
            bound.toDouble(), FluidCollisionMode.ALWAYS
        )
        return result?.hitBlock?.location
    }

    @JvmStatic
    fun goVelocity(entity: LivingEntity, target: Location, value: Double) {
        entity.velocity = entity.velocity.add(
            target.toVector()
                .subtract(entity.location.toVector()).normalize()
                .multiply(value)
        )
    }

    @JvmStatic
    fun piercingDamage(entity: LivingEntity, damage: Double) {
        entity.health = maxOf(0.0, entity.health - damage)
    }

    @JvmStatic
    fun healEntity(entity: LivingEntity, amount: Double) {
        val maxHealthValue = AttributeUtils.getMaxHealth(entity)
        entity.health = minOf(maxHealthValue, entity.health + amount)
    }

    @JvmStatic
    @JvmOverloads
    fun createBox(center: Location, material: Material, radius: Int, height: Int, ignoreBedrock: Boolean = false) {
        val world = center.world ?: return
        for (y in 1..height) {
            for (x in -radius..radius) {
                for (z in -radius..radius) {
                    val block = world.getBlockAt(center.clone()
                        .add(x.toDouble(), y.toDouble(), z.toDouble()))
                    if (ignoreBedrock || block.type != Material.BEDROCK) {
                        block.type = material
                    }
                }
            }
        }
    }

    @JvmStatic
    fun splashDamage(caster: Player, location: Location, bound: Double, damage: Double) {
        caster.world.getNearbyEntities(location, bound, bound, bound)
            .filterIsInstance<LivingEntity>()
            .filter { it !== caster && isDifferentTeam(caster, it) }
            .forEach { it.damage(damage, caster) }
    }

    @JvmStatic
    @JvmOverloads
    fun splashTask(
        caster: Player, location: Location, bound: Double,
        filter: Predicate<LivingEntity> = Predicate { true },
        action: Consumer<LivingEntity>
    ) {
        caster.world.getNearbyEntities(location, bound, bound, bound)
            .filterIsInstance<LivingEntity>()
            .filter { it !== caster && isDifferentTeam(caster, it) }
            .filter { filter.test(it) }
            .forEach { action.accept(it) }
    }

    private const val POINTS_PER_BLOCK = 4.0
    private const val INTERVAL_TICKS = 5L

    @JvmStatic
    fun createCircleIndicator(
        player: Player,
        radius: Double,
        particle: Particle,
    ): BukkitRunnable {
        val points = maxOf(8, (2 * Math.PI * radius * POINTS_PER_BLOCK).toInt())
        val task = object : BukkitRunnable() {
            override fun run() {
                if (!player.isOnline) {
                    cancel()
                    return
                }
                val center = player.location
                val world = center.world ?: return
                for (i in 0 until points) {
                    val angle = 2 * Math.PI * i / points
                    val x = center.x + radius * cos(angle)
                    val z = center.z + radius * sin(angle)
                    world.spawnParticle(particle, x, center.y + 0.5, z, 1, 0.0, 0.0, 0.0, 0.0)
                }
            }
        }
        task.runTaskTimer(PhysicalFighters.plugin, 0L, INTERVAL_TICKS)
        return task
    }

    private fun isDifferentTeam(caster: Player, target: LivingEntity): Boolean {
        if (target !is Player) return true
        val scoreboard = Bukkit.getScoreboardManager()?.mainScoreboard ?: return true
        val casterTeam = scoreboard.getEntryTeam(caster.name)
        val targetTeam = scoreboard.getEntryTeam(target.name)
        if (casterTeam == null || targetTeam == null) return true
        return casterTeam != targetTeam
    }
}